kasa/discover.py

```diff
@@ -378,20 +380,30 @@ def _get_device_instance(
                 f"Unable to read response from device: {ip}: {ex}"
             ) from ex
 
-        if (
-            discovery_result.mgt_encrypt_schm.encrypt_type == "KLAP"
-            and discovery_result.mgt_encrypt_schm.lv is None
-        ):
+        if discovery_result.mgt_encrypt_schm.encrypt_type in ("KLAP", "AES"):
             type_ = discovery_result.device_type
             device_class = None
-            if type_.upper() == "IOT.SMARTPLUGSWITCH":
-                device_class = SmartPlug
+            supported_device_types: dict[str, Type[SmartDevice]] = {
+                "SMART.TAPOPLUG": TapoPlug,
+                "SMART.KASAPLUG": TapoPlug,
+                "IOT.SMARTPLUGSWITCH": SmartPlug,
+            }
+
+            if discovery_result.device_type in supported_device_types:
+                device_class = supported_device_types[discovery_result.device_type]
 
             if device_class:
                 _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
                 device = device_class(ip, port=port, credentials=credentials)
                 device.update_from_discover_info(discovery_result.get_dict())
-                device.protocol = TPLinkKlap(ip, credentials=credentials)
+                if discovery_result.mgt_encrypt_schm.encrypt_type == "KLAP":
```

Comment:
Create a dict of supported supported types, use `get()` and log & raise an exception for unsupported ones.

Reply:
Done.  This method is a lot more readable now thanks!